internal/jsonwall/jsonwall.go

```diff
@@ -0,0 +1,307 @@
+		}
+	}
+	db.count = len(db.index) - 1 // Last item is end of buffer.
+	return db, nil
+}
+
+// DB holds an in-memory read-only database ready for querying.
+type DB struct {
+	schema string
+	data   []byte
+	index  []int
+	count  int
+}
+
+// Schema returns the optional schema value that was provided when writing the database.
+// This information is available to help with the decoding of the data, and is not
+// internally interpreted in any way.
+func (db *DB) Schema() string {
+	return db.schema
+}
+
+func (db *DB) prefix(value any) ([]byte, error) {
+	data, err := json.Marshal(value)
+	if err != nil || len(data) == 0 || data[0] != '{' {
+		return nil, fmt.Errorf("invalid database search value: %#v", value)
+	}
+	data[len(data)-1] = ','
+	return data, nil
+}
+
+var ErrNotFound = fmt.Errorf("value not found in database")
+
+const notFound = -2
+
+func (db *DB) search(prefix []byte) (i int, err error) {
+	i = sort.Search(db.count, func(i int) bool {
+		res := bytes.Compare(db.data[db.index[i]:], prefix) >= 0
+		return res
+	})
+	if db.match(i, prefix) {
+		return i, nil
+	}
+	return notFound, ErrNotFound
+}
+
+func (db *DB) iterate(prefix []byte) *Iterator {
+	i := sort.Search(db.count, func(i int) bool {
+		res := bytes.Compare(db.data[db.index[i]:], prefix) >= 0
+		return res
+	})
+	return &Iterator{db: db, prefix: prefix, pos: i, next: i}
+}
+
+func (db *DB) match(i int, prefix []byte) bool {
+	if i < 0 || i >= db.count {
+		return false
+	}
+	return bytes.HasPrefix(db.data[db.index[i]:], prefix)
+}
+
+func (db *DB) decode(i int, value any) error {
+	return json.Unmarshal(db.data[db.index[i]:db.index[i+1]], value)
+}
+
+// Get encodes the provided value as JSON, finds the first entry in the
+// database with initial fields exactly matching that encoding, and then
+// decodes the entry back into the provided value.
+func (db *DB) Get(value any) error {
+	prefix, err := db.prefix(value)
+	if err != nil {
+		return err
+	}
+	i, err := db.search(prefix)
+	if err != nil {
+		return err
+	}
+	return db.decode(i, value)
+}
+
+// Iterate encodes the provided value as JSON and returns an iterator that will
+// go over every entry in the database with initial fields that exactly match
+// that encoding. An iterator is still returned even if no entries match the
+// provided value.
+func (db *DB) Iterate(value any) (*Iterator, error) {
+	if value == nil {
+		return &Iterator{db: db}, nil
+	}
+	prefix, err := db.prefix(value)
+	if err != nil {
+		return nil, err
+	}
+	return db.iterate(prefix), nil
+}
+
+// IteratePrefix works similarly to Iterate, except that after encoding the
+// provided value as JSON, its last encoded field must be a string that will be
+// matched as a prefix of the respective database entry field, instead of being
+// matched exactly.
+func (db *DB) IteratePrefix(value any) (*Iterator, error) {
+	prefix, err := db.prefix(value)
+	if err != nil {
+		return nil, err
+	}
+	if !bytes.HasSuffix(prefix, []byte{'"', ','}) {
```

Comment:
I think we touched this in person already. Shouldn't this also include `"}` ending, i.e. something like the following?
```
!bytes.HasSuffix(prefix, []byte{'"', ','}) || !bytes.HasSuffix(prefix, []byte{'"', '}'})
```

Reply:
Indeed, thanks.

Reply:
Actually, we don't need this, because the prefix is being manufactured by the prefix method. The comma is forcefully injected there.